### ISSUE TITLE
:no-require prevents loading at init time

### ISSUE THREAD

**Issue by SwiftLawnGnome** [jwiegley/use-package#629]
If I understand the documentation correctly, `:no-require` should only prevent the package from being loaded at compile time. However,

``` emacs-lisp
(use-package package
  :no-require t
  :config (add-hook 'after-init-hook
                    (lambda () (message "'package' was loaded"))))
```
expands to

``` emacs-lisp
(progn
  (add-hook 'after-init-hook
	    (lambda nil
	      (message "'package' was loaded")))
  t)
```

Without the `:no-require`, it expands to

``` emacs-lisp
(progn
  (require 'package nil nil)
  (add-hook 'after-init-hook
	    (lambda nil
	      (message "'package' was loaded")))
  t)
```

There are a number of cases when this causes problems, for example when I want a package to load at init time, but only when its `:if` clause succeeds. The `:no-require` also makes `:demand t` do nothing.

**Comment by jwiegley** [comment]
Actually, the documentation is wrong.  All that `:no-require` does is to never emit a `require` statement, either at compile time or at load time.

**Comment by jwiegley** [comment]
I'm not clear yet on what actually you want to do. Can you describe in greater detail what you're trying to achieve? And even better, what you would have written to do it manually?

**Comment by SwiftLawnGnome** [comment]
Sure
I use EXWM about 50% of the time. Near the top of my init file, I have

``` emacs-lisp
(use-package my-exwm-config
  :when (getenv "EXWM")
  :config (add-hook 'after-init-hook (lambda () (setenv "EXWM"))))

```
I set `$EXWM` in my start-exwm.sh script, so that my EXWM settings (which includes the use-package declaration for EXWM itself) only get loaded when I need it. I also have a few other packages and settings that I only want loaded when I’m in EXWM, for example

``` emacs-lisp
(use-package fancy-battery
  :straight t
  :requires exwm
  :after my-spaceline
  :config (fancy-battery-mode 1))

```
(I use straight.el as my package manager)
I only need battery information in my mode-line when I’m in EXWM, but if I recompile my init file it will load EXWM, and then load fancy-battery because the `:requires` succeeds. When I changed the exwm declaration to

``` emacs-lisp
(when (getenv "EXWM")
  (require 'my-exwm-config nil nil)
  (add-hook 'after-init-hook
	    (lambda nil (setenv "EXWM"))))
```

(which is actually just the macro expansion of the use-package form) fancy-battery no longer got loaded. Not a major problem, and I also realized since my last comment that using `:no-require t` with `:config (require ‘this-package)` would do what I’d like, I believe. I still don’t fully understand how use-package processes and compiles each form, so maybe I’m missing something.
